components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -757,6 +766,695 @@ impl Interpolate for LengthOrNone {
+                    matrix.m3${i} += decomposed.translate.${i - 1} * matrix.m${j}${i};
+                % endfor
+            % endfor
+
+            // Apply rotation
+            let x = decomposed.quaternion.0;
+            let y = decomposed.quaternion.1;
+            let z = decomposed.quaternion.2;
+            let w = decomposed.quaternion.3;
+
+            // Construct a composite rotation matrix from the quaternion values
+            // rotationMatrix is a identity 4x4 matrix initially
+            let mut rotation_matrix = ComputedMatrix::identity();
+            rotation_matrix.m11 = 1.0 - 2.0 * (y * y + z * z);
+            rotation_matrix.m12 = 2.0 * (x * y - z * w);
+            rotation_matrix.m13 = 2.0 * (x * z + y * w);
+            rotation_matrix.m21 = 2.0 * (x * y + z * w);
+            rotation_matrix.m22 = 1.0 - 2.0 * (x * x + z * z);
+            rotation_matrix.m23 = 2.0 * (y * z - x * w);
+            rotation_matrix.m31 = 2.0 * (x * z - y * w);
+            rotation_matrix.m32 = 2.0 * (y * z + x * w);
+            rotation_matrix.m33 = 1.0 - 2.0 * (x * x + y * y);
+
+            matrix = multiply(matrix, rotation_matrix);
+
+            // Apply skew
+            let mut temp = ComputedMatrix::identity();
+            if decomposed.skew.2 != 0.0 {
+                temp.m32 = decomposed.skew.2;
+                matrix = multiply(matrix, temp);
+
+            }
+
+            if decomposed.skew.1 != 0.0 {
+                temp.m32 = 0.0;
+                temp.m31 = decomposed.skew.1;
+                matrix = multiply(matrix, temp);
+            }
+
+            if decomposed.skew.0 != 0.0 {
+                temp.m31 = 0.0;
+                temp.m21 = decomposed.skew.0;
+                matrix = multiply(matrix, temp);
+            }
+
+            // Apply scale
+            % for i in range(1, 4):
+                % for j in range(1, 4):
+                    matrix.m${i}${j} *= decomposed.scale.${i - 1};
+                % endfor
+            % endfor
+
+            matrix
+        }
+    }
+
+    // Multiplication of two 4x4 matrices.
+    fn multiply(a: ComputedMatrix, b: ComputedMatrix) -> ComputedMatrix {
+        let mut a_clone = a;
+        % for i in range(1, 5):
+            % for j in range(1, 5):
+                a_clone.m${i}${j} = (a.m${i}1 * b.m1${j}) +
+                                   (a.m${i}2 * b.m2${j}) +
+                                   (a.m${i}3 * b.m3${j}) +
+                                   (a.m${i}4 * b.m4${j});
+            % endfor
+        % endfor
+        a_clone
+    }
+
+    impl ComputedMatrix {
+        fn is_3d(&self) -> bool {
+            self.m13 != 0.0 || self.m14 != 0.0 ||
+            self.m23 != 0.0 || self.m24 != 0.0 ||
+            self.m31 != 0.0 || self.m32 != 0.0 || self.m33 != 1.0 || self.m34 != 0.0 ||
+            self.m43 != 0.0 || self.m44 != 1.0
+        }
+
+        pub fn determinant(&self) -> CSSFloat {
+            self.m14 * self.m23 * self.m32 * self.m41 -
```

Comment:
this could be mako generated, but not necessary.

It can be done by iterating over `for (x,y,z,w) itertools.permutations([1,2,3,4])`, multiplying the elements at index `0x*1y*2z*3w`, and adding a sign for the parity of the permutation. But this can get hairy too.